### PR TITLE
Type check bounds expression arguments

### PR DESCRIPTION
**Pull request by dtarditi**
This change adds type checking for bounds expressions of the form (bounds(e1, e2).  The type checking rules are:
- The types of e1 and e2 must both be pointer types.
- The types e1 and e2 must both be pointers to objects or incomplete types, and
- The types of e1 and e2 must both be pointers to the same type or one of them must be a pointer to void.

Testing: 
- New tests will be added to typechecking\bounds.c in the Checked C GitHub repo.  These will be committed separately.
- Passes existing clang tests.

### CODE REVIEW

File: lib/Sema/SemaExpr.cpp [hunk 1]
```diff
@@ -11949,14 +11949,67 @@ ExprResult Sema::ActOnCountBoundsExpr(SourceLocation BoundsKWLoc,
...
+  Result = UsualUnaryConversions(UpperBound);
+  if (Result.isInvalid())
+    return ExprError();
+  UpperBound = Result.get();
+
+  const Type *LowerBoundPointee = ValidateBoundsExprArgument(LowerBound);
+  const Type *UpperBoundPointee = ValidateBoundsExprArgument(UpperBound);
+
+  if (!LowerBoundPointee || !UpperBoundPointee)
+    return ExprError();
```
Comment (awruef) [inline lib/Sema/SemaExpr.cpp]: How could this happen? Should there be a Diag here too?

File: lib/Sema/SemaExpr.cpp [hunk 2]
```diff
@@ -11949,14 +11949,67 @@ ExprResult Sema::ActOnCountBoundsExpr(SourceLocation BoundsKWLoc,
...
 ExprResult Sema::ActOnRangeBoundsExpr(SourceLocation BoundsKWLoc,
-                                            Expr *LowerBound,
-                                            Expr *UpperBound,
-                                            SourceLocation RParenLoc) {
+                                      Expr *LowerBound,
+                                      Expr *UpperBound,
+                                      SourceLocation RParenLoc) {
+  ExprResult Result = UsualUnaryConversions(LowerBound);
+  if (Result.isInvalid())
+     return ExprError();
```
Comment (awruef) [inline lib/Sema/SemaExpr.cpp]: How could this happen? Should there be a Diag here too?

**Comment by awruef** [comment]
Left two comments, otherwise looks fine to me.

File: lib/Sema/SemaExpr.cpp [hunk 3]
```diff
@@ -11949,14 +11949,67 @@ ExprResult Sema::ActOnCountBoundsExpr(SourceLocation BoundsKWLoc,
...
+  return new (Context) CountBoundsExpr(Kind, CountExpr, BoundsKWLoc,
                                        RParenLoc);
 }
 
+/// \brief Validate the type of an argument expression to a bounds
+/// expression. If valid, return the unqualified pointee type.  Otherwise
+/// return null.
+const Type *Sema::ValidateBoundsExprArgument(Expr *Arg) {
+  QualType ArgType = Arg->getType();
+  const Type *ArgPointerType = ArgType->getAs<PointerType>();
```
Comment (reubeno) [inline lib/Sema/SemaExpr.cpp]: >PointerType [](start = 46, length = 11)

Naive qn: does this rule out array types, e.g.: an int[]?

File: lib/Sema/SemaExpr.cpp [hunk 4]
```diff
@@ -11949,14 +11949,67 @@ ExprResult Sema::ActOnCountBoundsExpr(SourceLocation BoundsKWLoc,
...
+  if (!ArgPointerType) {
+    Diag(Arg->getLocStart(),
+         diag::err_typecheck_pointer_type_expected) << ArgType;
+    return nullptr;
+  }
+  QualType ArgTypePointee = ArgPointerType->getPointeeType();
+  // For the Checked C extension, we don't expect to see of the clang
```
Comment (reubeno) [inline lib/Sema/SemaExpr.cpp]: >of [](start = 57, length = 2)

[Nit] Missing word in the comment?

File: lib/Sema/SemaExpr.cpp [hunk 5]
```diff
@@ -11949,14 +11949,67 @@ ExprResult Sema::ActOnCountBoundsExpr(SourceLocation BoundsKWLoc,
...
+  // extensions for qualifiers.  Assert if we do.
+  Qualifiers ArgTypeQuals = ArgTypePointee.getQualifiers();
+  ArgTypeQuals.removeCVRQualifiers();
+  assert(!ArgTypeQuals.hasQualifiers() &&
+         "unexpected non-CVR qualifiers on type");
+
+  if (!ArgTypePointee->isIncompleteOrObjectType()) {
+    Diag(Arg->getLocStart(), diag::err_typecheck_bounds_expr) << ArgType;
+   return nullptr;
```
Comment (reubeno) [inline lib/Sema/SemaExpr.cpp]: >    [](start = 0, length = 3)

[Nit] Spacing looks to be off by a char.

File: lib/Sema/SemaExpr.cpp [hunk 6]
```diff
@@ -11949,14 +11949,67 @@ ExprResult Sema::ActOnCountBoundsExpr(SourceLocation BoundsKWLoc,
...
+  QualType ArgType = Arg->getType();
+  const Type *ArgPointerType = ArgType->getAs<PointerType>();
+  if (!ArgPointerType) {
+    Diag(Arg->getLocStart(),
+         diag::err_typecheck_pointer_type_expected) << ArgType;
+    return nullptr;
+  }
+  QualType ArgTypePointee = ArgPointerType->getPointeeType();
+  // For the Checked C extension, we don't expect to see of the clang
+  // extensions for qualifiers.  Assert if we do.
```
Comment (reubeno) [inline lib/Sema/SemaExpr.cpp]: >extensions for qualifiers [](start = 5, length = 25)

Out of curiosity, does clang provide support for some of the Microsoft C/C++ pointer extensions (e.g. __sptr/__uptr, __ptr32/__ptr64, __unaligned)?  If so, do they show up as qualifiers or in some other way?  Will they get in our way?

**Comment by reubeno** [comment]
Ditto.  Left a few questions, otherwise looks good.

File: lib/Sema/SemaExpr.cpp (hunk 1 above)
Comment (dtarditi) [inline lib/Sema/SemaExpr.cpp]: For the C path, there appear to a few relatively obscure (but valid) ways in which UsualUnaryConversions can return an invalid ExprResult.   We don't need diagnostics here.  The convention in clang is that the function that returns an invalid ExprResult should emit any necessary diagnostic.

File: lib/Sema/SemaExpr.cpp (hunk 2 above)
Comment (dtarditi) [inline lib/Sema/SemaExpr.cpp]: See the comment below.

File: lib/Sema/SemaExpr.cpp (hunk 5 above)
Comment (dtarditi) [inline lib/Sema/SemaExpr.cpp]: Will fix this.

File: lib/Sema/SemaExpr.cpp (hunk 6 above)
Comment (dtarditi) [inline lib/Sema/SemaExpr.cpp]: They do support them, but they are treated as attributes on types, not modifiers.    I'll add a longer comment explaining what is going on with clang extensions to type qualifiers.

File: lib/Sema/SemaExpr.cpp (hunk 3 above)
Comment (dtarditi) [inline lib/Sema/SemaExpr.cpp]: Good question.  The function assumes that array->pointer conversion has already been done, so it does not rule out array types.   I've added a comment to the header for the function noting the assumption.

**Comment by dtarditi** [comment]
I've address the code review feedback and improved commenting. I also answered the questions above.  I'll commit this now.  If you want additional changes, let me know.

**Pull request merged**